capa/features/extractors/dnfile/extractor.py

```diff
@@ -19,15 +20,44 @@
 import capa.features.extractors.dnfile.function
 from capa.features.common import Feature
 from capa.features.address import NO_ADDRESS, Address, DNTokenAddress, DNTokenOffsetAddress
+from capa.features.extractors.dnfile.types import DnfileFeatureCache, DnfileFeatureExtractorType
 from capa.features.extractors.base_extractor import BBHandle, InsnHandle, FunctionHandle, FeatureExtractor
-from capa.features.extractors.dnfile.helpers import get_dotnet_managed_method_bodies
+from capa.features.extractors.dnfile.helpers import (
+    get_dotnet_types,
+    get_dotnet_fields,
+    get_dotnet_managed_imports,
+    get_dotnet_managed_methods,
+    get_dotnet_unmanaged_imports,
+    get_dotnet_managed_method_bodies,
+)
 
 
 class DnfileFeatureExtractor(FeatureExtractor):
     def __init__(self, path: str):
         super().__init__()
         self.pe: dnfile.dnPE = dnfile.dnPE(path)
 
+        # cached .NET token lookup tables
+        self.tokens: Dict[DnfileFeatureCache, Dict[int, DnfileFeatureExtractorType]] = {}
+        for value in DnfileFeatureCache:
+            self.tokens[value] = {}
```

Comment:
Use a defaultdict or a new class here instead?

Reply:
looking at this below it seems complex, would it make sense to have a `Cache` object that we can get from (instead of `get_cached_token`) etc.?

Reply:
good point; updated